lib/src/kbasesearchengine/main/NarrativeInfoDecorator.java

```diff
@@ -96,51 +105,64 @@ public SearchObjectsOutput searchObjects(final SearchObjectsInput params, final
                     params.getPostProcessing().getAddNarrativeInfo() == 1) {
                 getObjsOutput = getObjsOutput.withAccessGroupNarrativeInfo(addNarrativeInfo(
                         getObjsOutput.getObjects(),
-                        getObjsOutput.getAccessGroupNarrativeInfo()));
+                        getObjsOutput.getAccessGroupNarrativeInfo(),
+                        inputArr));
             }
         }
         return getObjsOutput;
     }
 
+    /**
+     * Adds narrative information for non deleted and valid narrative workspaces. Removes results otherwise and
+     * log list of removed guids. If env "KBASE_SEARCH_SHOW_REMOVED_GUIDS", is set to true, list of removed guids
+     * is added to additionalProperties.
+     */
     private Map<Long, Tuple5 <String, Long, Long, String, String>> addNarrativeInfo(
             final List<ObjectData> objects,
-            final Map<Long, Tuple5 <String, Long, Long, String, String>> accessGroupNarrInfo) {
+            final Map<Long, Tuple5 <String, Long, Long, String, String>> accessGroupNarrInfo,
+            final List<String> removedGuids) {
 
         final Map<Long, Tuple5 <String, Long, Long, String, String>> retVal = new HashMap<>();
 
         if (accessGroupNarrInfo != null) {
             retVal.putAll(accessGroupNarrInfo);
         }
-        final Set<Long> wsIdsSet = new HashSet<>();
-
-        for (final ObjectData objData: objects) {
+        final Set<String> userNames = new HashSet<>();
+        final Iterator<ObjectData> iter = objects.iterator();
+        while (iter.hasNext()) {
+            final ObjectData objData = iter.next();
             final GUID guid = new GUID(objData.getGuid());
             if (WorkspaceEventHandler.STORAGE_CODE.equals(guid.getStorageCode())) {
-                wsIdsSet.add((long) guid.getAccessGroupId());
-            }
-        }
-        final Set<String> userNames = new HashSet<>();
-        for (final long workspaceId: wsIdsSet) {
-            final NarrativeInfo narrInfo = narrInfoProvider.findNarrativeInfo(workspaceId);
-            if (narrInfo != null) {
-                final Tuple5<String, Long, Long, String, String> tempNarrInfo =
-                        new Tuple5<String, Long, Long, String, String>()
-                                .withE1(narrInfo.getNarrativeName())
-                                .withE2(narrInfo.getNarrativeId())
-                                .withE3(narrInfo.getTimeLastSaved())
-                                .withE4(narrInfo.getWsOwnerUsername());
-                userNames.add(tempNarrInfo.getE4());
-                retVal.put(workspaceId, tempNarrInfo);
-            }
-            else {
-                retVal.put(workspaceId, null);
+                final long workspaceId = guid.getAccessGroupId();
+                final NarrativeInfo narrInfo = narrInfoProvider.findNarrativeInfo(workspaceId);
+
+                if (narrInfo != null) {
```

Comment:
Comment here noting that narrInfo is only null if the workspace is deleted. If that's not documented in the provider class should add docs in the PR that fixes the bug there.

Reply:
what do do if it is a non deletion exception then? Right now that also returns null, but idk what we would want to change that too

Reply:
If it's not a deletion exception then the exception should be thrown (or possibly wrapped) and thrown, as the deleted exception is the only exception we should expect under normal circumstances, AFAIK